# Breast Cancer EDA & Modeling Notebook

## Imports & Global Config

## Quick Guide: Running Manual vs Agentic EDA

This notebook contains both a manual EDA workflow and an automated `EDAAgent` for a friendly comparison. Follow these steps to reproduce the demo:

1. Install dependencies: `pip install -r requirements.txt` (see repository root).
2. Open this notebook in Jupyter and run cells top-to-bottom to ensure all variables are initialized.
3. Manual EDA: run the cells up through the 'Manual EDA Summary' section for the human-driven exploration and visualizations.
4. Agentic EDA: run the cell in the 'Running the EDA Agent' section to execute `EDAAgent` — watch the timestamped logs it prints.
5. Agentic FE demo: a short, non-redundant 'Agentic-style Feature Engineering Demo' appears just before 'Baseline Feature Engineering' — it reuses existing helper functions to demonstrate reproducible FE in one compact step.

For professor review: show the Agent Execution Log and the Comparison section (Manual vs. Agentic) — they demonstrate reproducibility, auditability, and parity with the manual workflow.

If you need a clean environment, use the included `requirements.txt` and a venv. See the project root for instructions.

In [1]:
# Basic libraries
import os
import sys
import time
import logging
import importlib.util
import inspect
import subprocess
from pathlib import Path
from datetime import datetime

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import mutual_info_classif
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.metrics import roc_auc_score, classification_report, ConfusionMatrixDisplay
import joblib
from xgboost import XGBClassifier

# Display + plotting style
pd.set_option("display.max_columns", 200)
sns.set(style="whitegrid")

# For reproducibility
RANDOM_STATE = 42


## Repository Root & Project Paths
This cell detects the project’s root folder and sets up all directory paths (data, engineered data, artifacts) so the notebook can reliably read and save files no matter where it’s run.


In [2]:
# Find the project root so all file paths work no matter where the notebook is run
def get_repo_root():
    try:
        # Ask Git for the top-level directory of this repository
        return Path(subprocess.check_output(["git", "rev-parse", "--show-toplevel"], text=True).strip())
    except Exception:
        # Walk up a few levels looking for a .git folder
        p = Path.cwd()
        for _ in range(6):
            if (p / ".git").exists():
                return p
            p = p.parent
        # Fallback: current working directory
        return Path.cwd()

try:
    ROOT = get_repo_root()
except Exception:
    ROOT = Path.cwd()

FALLBACK_ROOT = Path(r"C:\Users\rajni\Documents\breast-cancer-agentic")
if ROOT != FALLBACK_ROOT and not (ROOT / "data" / "raw").exists():
    ROOT = FALLBACK_ROOT

print("Repo root:", ROOT)

# Define key directories
DATA_RAW        = ROOT / "data" / "raw"
DATA_ENGINEERED = ROOT / "data" / "engineered"
ARTIFACTS_ENG   = ROOT / "artifacts" / "engineering"
ARTIFACTS_EDA   = ROOT / "artifacts" / "eda"

for p in [DATA_ENGINEERED, ARTIFACTS_ENG, ARTIFACTS_EDA]:
    p.mkdir(parents=True, exist_ok=True)

# Source CSV + target column
SRC_FILE   = DATA_RAW / "breast_cancer_with_columns.csv"
TARGET_COL = "diagnosis"

print("Using source file:", SRC_FILE)
assert SRC_FILE.exists(), f"Missing source file: {SRC_FILE}"


Repo root: C:\Users\rajni\Documents\breast-cancer-agentic
Using source file: C:\Users\rajni\Documents\breast-cancer-agentic\data\raw\breast_cancer_with_columns.csv


## Load Dataset & Basic Checks
This cell loads the raw dataset, verifies that the target column exists, splits the data into features and labels, and provides basic previews (head, summary stats, missing values) to confirm the data was loaded correctly.

In [3]:
# Read the dataset from disk
df = pd.read_csv(SRC_FILE)
print("Loaded shape:", df.shape)

# Make sure the target column exists before continuing
if TARGET_COL not in df.columns:
    print("Columns available (first 40):", list(df.columns)[:40])
    raise AssertionError(f"Target column '{TARGET_COL}' not found in CSV")

# Separate features (X) from the target (y)
y = df[TARGET_COL]
X = df.drop(columns=[TARGET_COL])

print("Target distribution (normalized):")
print(y.value_counts(normalize=True))

# Glance at the data and summary statistics to sanity-check values and types
display(df.head())
display(df.describe(include="all").T)

print("Missing values (top 15):")
print(df.isna().sum().sort_values(ascending=False).head(15))


Loaded shape: (569, 32)
Target distribution (normalized):
diagnosis
B    0.627417
M    0.372583
Name: proportion, dtype: float64


id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  fractal_dimension_mean  radius_se  texture_se  perimeter_se  \
0         0.2419                 0.07871     1.0950      0.9053         8.589   
1         0.1812                 0.05667     0.5435      0.7339         3.398   
2         0.2069                 0.05999     0.7456      0.7869         4.585   
3         0.2597                 0.09744     0.4956      1.1560         3.445   
4         0.1809                 0.05883     0.7572      0.7813         5.438   

   area_se  smoothness_se  compactness_se  concavity_se  concave points_se  \
0   153.40       0.006399         0.04904       0.05373            0.01587   
1    74.08       0.005225         0.01308       0.01860            0.01340   
2    94.03       0.006150         0.04006       0.03832            0.02058   
3    27.23       0.009110         0.07458       0.05661            0.01867   
4    94.44       0.011490         0.02461       0.05688            0.01885   

   symmetry_se  fractal_dimension_se  radius_worst  texture_worst  \
0      0.03003              0.006193         25.38          17.33   
1      0.01389              0.003532         24.99          23.41   
2      0.02250              0.004571         23.57          25.53   
3      0.05963              0.009208         14.91          26.50   
4      0.01756              0.005115         22.54          16.67   

   perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
0           184.60      2019.0            0.1622             0.6656   
1           158.80      1956.0            0.1238             0.1866   
2           152.50      1709.0            0.1444             0.4245   
3            98.87       567.7            0.2098             0.8663   
4           152.20      1575.0            0.1374             0.2050   

   concavity_worst  concave points_worst  symmetry_worst  \
0           0.7119                0.2654          0.4601   
1           0.2416                0.1860          0.2750   
2           0.4504                0.2430          0.3613   
3           0.6869                0.2575          0.6638   
4           0.4000                0.1625          0.2364   

   fractal_dimension_worst  
0                  0.11890  
1                  0.08902  
2                  0.08758  
3                  0.17300  
4                  0.07678

count unique  top freq             mean  \
id                       569.0    NaN  NaN  NaN  30371831.432337   
diagnosis                  569      2    B  357              NaN   
radius_mean              569.0    NaN  NaN  NaN        14.127292   
texture_mean             569.0    NaN  NaN  NaN        19.289649   
perimeter_mean           569.0    NaN  NaN  NaN        91.969033   
area_mean                569.0    NaN  NaN  NaN       654.889104   
smoothness_mean          569.0    NaN  NaN  NaN          0.09636   
compactness_mean         569.0    NaN  NaN  NaN         0.104341   
concavity_mean           569.0    NaN  NaN  NaN         0.088799   
concave points_mean      569.0    NaN  NaN  NaN         0.048919   
symmetry_mean            569.0    NaN  NaN  NaN         0.181162   
fractal_dimension_mean   569.0    NaN  NaN  NaN         0.062798   
radius_se                569.0    NaN  NaN  NaN         0.405172   
texture_se               569.0    NaN  NaN  NaN         1.216853   
perimeter_se             569.0    NaN  NaN  NaN         2.866059   
area_se                  569.0    NaN  NaN  NaN        40.337079   
smoothness_se            569.0    NaN  NaN  NaN         0.007041   
compactness_se           569.0    NaN  NaN  NaN         0.025478   
concavity_se             569.0    NaN  NaN  NaN         0.031894   
concave points_se        569.0    NaN  NaN  NaN         0.011796   
symmetry_se              569.0    NaN  NaN  NaN         0.020542   
fractal_dimension_se     569.0    NaN  NaN  NaN         0.003795   
radius_worst             569.0    NaN  NaN  NaN         16.26919   
texture_worst            569.0    NaN  NaN  NaN        25.677223   
perimeter_worst          569.0    NaN  NaN  NaN       107.261213   
area_worst               569.0    NaN  NaN  NaN       880.583128   
smoothness_worst         569.0    NaN  NaN  NaN         0.132369   
compactness_worst        569.0    NaN  NaN  NaN         0.254265   
concavity_worst          569.0    NaN  NaN  NaN         0.272188   
concave points_worst     569.0    NaN  NaN  NaN         0.114606   
symmetry_worst           569.0    NaN  NaN  NaN         0.290076   
fractal_dimension_worst  569.0    NaN  NaN  NaN         0.083946   

                                      std       min       25%       50%  \
id                       125020585.612224    8670.0  869218.0  906024.0   
diagnosis                             NaN       NaN       NaN       NaN   
radius_mean                      3.524049     6.981      11.7     13.37   
texture_mean                     4.301036      9.71     16.17     18.84   
perimeter_mean                  24.298981     43.79     75.17     86.24   
area_mean                      351.914129     143.5     420.3     551.1   
smoothness_mean                  0.014064   0.05263   0.08637   0.09587   
compactness_mean                 0.052813   0.01938   0.06492   0.09263   
concavity_mean                    0.07972       0.0   0.02956   0.06154   
concave points_mean              0.038803       0.0   0.02031    0.0335   
symmetry_mean                    0.027414     0.106    0.1619    0.1792   
fractal_dimension_mean            0.00706   0.04996    0.0577   0.06154   
radius_se                        0.277313    0.1115    0.2324    0.3242   
texture_se                       0.551648    0.3602    0.8339     1.108   
perimeter_se                     2.021855     0.757     1.606     2.287   
area_se                         45.491006     6.802     17.85     24.53   
smoothness_se                    0.003003  0.001713  0.005169   0.00638   
compactness_se                   0.017908  0.002252   0.01308   0.02045   
concavity_se                     0.030186       0.0   0.01509   0.02589   
concave points_se                 0.00617       0.0  0.007638   0.01093   
symmetry_se                      0.008266  0.007882   0.01516   0.01873   
fractal_dimension_se             0.002646  0.000895  0.002248  0.003187   
radius_worst                     4.833242      7.93    

Missing values (top 15):
id                        0
diagnosis                 0
radius_mean               0
texture_mean              0
perimeter_mean            0
area_mean                 0
smoothness_mean           0
compactness_mean          0
concavity_mean            0
concave points_mean       0
symmetry_mean             0
fractal_dimension_mean    0
radius_se                 0
texture_se                0
perimeter_se              0
dtype: int64


## Data Quality Report
This cell provides a comprehensive summary of data quality: missing value rates, duplicate rows, data type distribution, and numeric ranges. These metrics help identify potential data issues before modeling and inform cleaning strategies.


In [4]:
# Missing values summary and percentage
missing_summary = pd.DataFrame({
    'column': df.columns,
    'n_missing': [df[c].isna().sum() for c in df.columns],
    'pct_missing': [100 * df[c].isna().sum() / len(df) for c in df.columns],
    'dtype': df.dtypes.values,
})
missing_summary = missing_summary[missing_summary['n_missing'] > 0].sort_values('n_missing', ascending=False)

print("=" * 80)
print("DATA QUALITY REPORT")
print("=" * 80)
print(f"\nDataset Shape: {df.shape[0]} rows × {df.shape[1]} columns")

# Check for duplicates
n_duplicates = df.duplicated().sum()
print(f"Duplicate Rows: {n_duplicates} ({100 * n_duplicates / len(df):.2f}%)")

# Missing values
if len(missing_summary) > 0:
    print(f"\nMissing Values (columns with missing data):")
    display(missing_summary[['column', 'n_missing', 'pct_missing', 'dtype']])
else:
    print(f"\n No missing values detected.")

# Data type distribution
print(f"\nData Type Distribution:")
dtype_counts = df.dtypes.value_counts()
for dtype, count in dtype_counts.items():
    print(f"  {dtype}: {count} columns")

# Numeric column ranges
print(f"\nNumeric Column Ranges (min, max, mean):")
numeric_ranges = []
for col in df.select_dtypes(include=[np.number]).columns:
    numeric_ranges.append({
        'column': col,
        'min': df[col].min(),
        'max': df[col].max(),
        'mean': df[col].mean(),
        'std': df[col].std(),
    })
numeric_df = pd.DataFrame(numeric_ranges)
display(numeric_df.head(10))

print(f"\n" + "=" * 80)
print("Data quality checks complete. Ready for EDA and modeling.")
print("=" * 80)

DATA QUALITY REPORT

Dataset Shape: 569 rows × 32 columns
Duplicate Rows: 0 (0.00%)

 No missing values detected.

Data Type Distribution:
  float64: 30 columns
  int64: 1 columns
  object: 1 columns

Numeric Column Ranges (min, max, mean):


column         min           max          mean           std
0                   id  8670.00000  9.113205e+08  3.037183e+07  1.250206e+08
1          radius_mean     6.98100  2.811000e+01  1.412729e+01  3.524049e+00
2         texture_mean     9.71000  3.928000e+01  1.928965e+01  4.301036e+00
3       perimeter_mean    43.79000  1.885000e+02  9.196903e+01  2.429898e+01
4            area_mean   143.50000  2.501000e+03  6.548891e+02  3.519141e+02
5      smoothness_mean     0.05263  1.634000e-01  9.636028e-02  1.406413e-02
6     compactness_mean     0.01938  3.454000e-01  1.043410e-01  5.281276e-02
7       concavity_mean     0.00000  4.268000e-01  8.879932e-02  7.971981e-02
8  concave points_mean     0.00000  2.012000e-01  4.891915e-02  3.880284e-02
9        symmetry_mean     0.10600  3.040000e-01  1.811619e-01  2.741428e-02


Data quality checks complete. Ready for EDA and modeling.


## Statistical Significance Testing
This cell performs t-tests and chi-square tests to quantify whether observed feature differences between classes are statistically significant (beyond random chance). Results inform feature selection and help flag spurious correlations.


In [5]:
!pip install statsmodels

In [6]:
from scipy import stats

# Perform t-tests to assess feature significance vs. target
# Separate feature values by class
class_0 = y_bin.unique()[0]
class_1 = y_bin.unique()[1]

X_class_0 = X[y_bin == class_0]
X_class_1 = X[y_bin == class_1]

# Compute t-stats and p-values for each numeric feature
test_results = []
for col in X.select_dtypes(include=[np.number]).columns:
    t_stat, p_val = stats.ttest_ind(X_class_0[col].dropna(), X_class_1[col].dropna())
    test_results.append({
        'feature': col,
        't_statistic': t_stat,
        'p_value': p_val,
        'significant': 'Yes' if p_val < 0.05 else 'No',
        'mean_class_0': X_class_0[col].mean(),
        'mean_class_1': X_class_1[col].mean(),
    })

significance_df = pd.DataFrame(test_results).sort_values('p_value')
print("Features ranked by statistical significance (t-test, α=0.05):")
display(significance_df.head(15))

# Log summary
n_significant = (significance_df['p_value'] < 0.05).sum()
print(f"\nTotal numeric features: {len(X.select_dtypes(include=[np.number]).columns)}")
print(f"Statistically significant features (p < 0.05): {n_significant}")



NameError: name 'y_bin' is not defined

## Numeric Overview & KDE Plots
This cell reviews all numeric columns, checks missing/unique counts, and generates KDE plots to visualize how feature distributions differ between malignant and benign cases.

In [ ]:
# Quick numeric overview and sample KDE plots
num_cols = [c for c in X.columns if pd.api.types.is_numeric_dtype(X[c])]
print("Numeric columns count:", len(num_cols))
display(pd.DataFrame({
    "col": num_cols,
    "n_missing": [X[c].isna().sum() for c in num_cols],
    "n_unique": [X[c].nunique() for c in num_cols],
}).sort_values(["n_missing", "n_unique"], ascending=[False, True]).head(20))

# Plot a few features
sel = num_cols[:6]
for col in sel:
    plt.figure(figsize=(6, 3))
    sns.kdeplot(data=df, x=col, hue=TARGET_COL, fill=True, common_norm=False)
    plt.title(f"{col} by {TARGET_COL}")
    plt.tight_layout()
    plt.show()


## Correlation Heatmap
This heatmap shows pairwise correlations between all numeric features.
It helps identify groups of highly related features and potential redundancy
before modeling.

In [ ]:
# Correlation Heatmap for All 30 Features
plt.figure(figsize=(10,8))
# Compute correlation matrix
corr = X.corr()
# Heatmap visualization
sns.heatmap(corr, cmap='coolwarm', center=0)
plt.title('Feature Correlation Heatmap')
plt.show()

## Outlier Detection & Visualization
This cell identifies univariate outliers using the IQR method (values beyond 1.5×IQR) and flags extreme values. This helps understand what gets clipped during percentile capping and validates the choice of clip thresholds.


In [ ]:
# OUTLIER SUMMARY USING IQR

# Create a list to store outlier information for each feature
outlier_summary = []

# Loop through all numeric columns in X
for col in X.select_dtypes(include=[np.number]).columns:
    
    # Compute the 25th and 75th percentiles
    Q1, Q3 = X[col].quantile([0.25, 0.75])
    
    # Calculate Interquartile Range (IQR)
    IQR = Q3 - Q1
    
    # Define lower and upper bounds for outliers
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    # Identify points outside the IQR bounds
    outliers = X[(X[col] < lower) | (X[col] > upper)][col]

    # Store results for this feature
    outlier_summary.append({
        "feature": col,
        "n_outliers": len(outliers),                    # Number of outlier values
        "pct_outliers": round(100 * len(outliers) / len(X), 2),   # Outlier percentage
        "lower_bound": lower,
        "upper_bound": upper,
        "min_val": X[col].min(),
        "max_val": X[col].max(),
    })

# Create a summary DataFrame and sort by number of outliers
outlier_df = pd.DataFrame(outlier_summary).sort_values("n_outliers", ascending=False)

print("Outlier Summary (IQR method):")
display(outlier_df[outlier_df["n_outliers"] > 0].head(10))


# BOXPLOTS FOR TOP 6 FEATURES WITH THE MOST OUTLIERS

# Select the top 6 features containing the most outliers
top_cols = outlier_df[outlier_df["n_outliers"] > 0]["feature"].head(6).tolist()

if top_cols:
    
    # Set up a 2x3 grid for plots
    fig, axes = plt.subplots(2, 3, figsize=(15, 8))
    axes = axes.flatten()

    # Create boxplots for each selected feature
    for idx, col in enumerate(top_cols):
        
        # Draw boxplot comparing class 0 and class 1
        axes[idx].boxplot(
            [X_class_0[col].dropna(), X_class_1[col].dropna()],
            labels=[f"Class {class_0}", f"Class {class_1}"]
        )
        
        # Get outlier count for plot title
        count = outlier_df.loc[outlier_df["feature"] == col, "n_outliers"].values[0]
        
        # Add feature name and outlier count
        axes[idx].set_title(f"{col} ({count} outliers)")
        axes[idx].set_ylabel("Value")

    # Adjust layout
    plt.tight_layout()
    
    # Display the boxplots
    plt.show()

print("\nNote: 1st–99th percentile capping can be applied to reduce the impact of these outliers.")


## Target Distribution Plot
This cell creates a simple countplot of the diagnosis column to show the class distribution (malignant vs. benign) and verify the dataset’s balance.

In [ ]:
plt.figure(figsize=(6, 4))
sns.countplot(x=TARGET_COL, data=df,
    palette=["#FF6B6B", "#4D96FF"]   # red (malignant), blue (benign)
)
plt.title("Distribution of Diagnosis")
plt.xlabel("Diagnosis")
plt.ylabel("Count")
plt.show()


## Class Imbalance Analysis
This cell checks whether the target classes are balanced. Imbalanced datasets may require resampling strategies (oversampling, undersampling, or class weights) during modeling to prevent the model from biasing toward the majority class.


In [ ]:
# Class Imbalance Check + Visuals

# Compute counts & percentages
counts = y_bin.value_counts().sort_index()
pcts = (counts / len(y_bin) * 100).round(2)
ratio = round(counts.max() / counts.min(), 2)

print("Class Distribution Summary:")
print(f"  Class 0: {counts[0]} samples ({pcts[0]}%)")
print(f"  Class 1: {counts[1]} samples ({pcts[1]}%)")
print(f"\nImbalance Ratio: {ratio}:1")

# Recommendation
if ratio > 1.5:
    print("\n Moderate class imbalance detected.")
    print("   • Consider class_weight='balanced'")
    print("   • SMOTE/oversampling are options")
    print("   • Use stratified CV")
    print("   • Track precision, recall, F1, ROC-AUC")
else:
    print("\n Classes are fairly balanced.")

# Visuals
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))

# Bar chart
ax1.bar(['Class 0', 'Class 1'], counts.values, color=['#4D96FF', '#FF6B6B'])
ax1.set_title("Class Count")
ax1.set_ylabel("Samples")
plt.savefig("class_distribution.png", dpi=300, bbox_inches="tight")
# Pie chart
ax2.pie(counts.values, labels=['Class 0', 'Class 1'], autopct='%1.1f%%',
        colors=['#4D96FF', '#FF6B6B'], startangle=90)
ax2.set_title("Class Percentage")

plt.tight_layout()
plt.show()


### Scatter, Pairplot, and PCA Visuals

This cell creates a few additional visualizations to explore how features relate
to the target:

- Scatter plot of the top two correlated features  
- Pairplot for a small set of top features  
- PCA 2D projection to visualize class separability  

These help show patterns in the data before modeling.


In [ ]:
# Simple scatter and PCA plots 
# 1) Scatter of two top correlated features
# 2) Small pairplot for top features
# 3) PCA 2D projection of the feature matrix

# Ensure we have a binary target variable to color plots
if 'y_bin' in globals():
    ybin = y_bin
else:
    if y.dtype == 'O':
        ybin = y.map(lambda v: 1 if str(v).lower().startswith('m') else 0)
    else:
        ybin = y

# Compute simple absolute correlations with target and pick top features
num_X = X.select_dtypes(include=[np.number])
corr_with_target = num_X.corrwith(ybin).abs().sort_values(ascending=False)
top_feats = corr_with_target.index.tolist()
print('Top features (by abs corr):', top_feats[:6])

# Scatter of two strongest features (if available)
if len(top_feats) >= 2:
    f1, f2 = top_feats[0], top_feats[1]
    plt.figure(figsize=(6,5))
    sns.scatterplot(data=df, x=f1, y=f2, hue=TARGET_COL, palette='Set1', alpha=0.7)
    plt.title(f'{f1} vs {f2} — colored by {TARGET_COL}')
    plt.xlabel(f1)
    plt.ylabel(f2)
    plt.tight_layout()
    plt.show()
else:
    print('Not enough numeric features for a two-feature scatter.')

# Pairplot for a small subset (top 3 or 4 features)
pair_feats = top_feats[:4] if len(top_feats) >= 2 else []
if len(pair_feats) >= 2:
    print('Drawing pairplot for:', pair_feats)
    sns.pairplot(df[pair_feats + [TARGET_COL]], hue=TARGET_COL, diag_kind='kde', plot_kws={'alpha':0.6})
    plt.suptitle('Pairplot — top correlated features', y=1.02)
    plt.show()

# PCA 2D projection (use scaled features if available)
from sklearn.decomposition import PCA

if 'X_scaled' in globals():
    X_pca_in = X_scaled.copy()
else:
    # Start from numeric columns, coerce any non-numeric entries to NaN
    X_pca_in = num_X.copy().apply(pd.to_numeric, errors='coerce')

# Replace infinite values with NaN
X_pca_in = X_pca_in.replace([np.inf, -np.inf], np.nan)

# Drop columns that are entirely NaN (they provide no signal)
all_nan_cols = X_pca_in.columns[X_pca_in.isna().all()].tolist()
if all_nan_cols:
    print("Dropping all-NaN columns before PCA:", all_nan_cols)
    X_pca_in = X_pca_in.drop(columns=all_nan_cols)

# Impute remaining NaNs with column mean
X_pca_in = X_pca_in.fillna(X_pca_in.mean())

# Final sanity check
assert not X_pca_in.isna().any().any(), "X_pca_in still contains NaN values after imputation"

pca = PCA(n_components=2, random_state=RANDOM_STATE)
X2 = pca.fit_transform(X_pca_in)

explained = pca.explained_variance_ratio_
plt.figure(figsize=(7,5))
sns.scatterplot(x=X2[:,0], y=X2[:,1], hue=ybin, palette='Set1', alpha=0.8)
plt.xlabel(f'PC1 ({explained[0]*100:.1f}% var)')
plt.ylabel(f'PC2 ({explained[1]*100:.1f}% var)')
plt.title('PCA 2D — features projected to 2D (colored by target)')
plt.tight_layout()
plt.show()

---
## Manual EDA Summary — What We've Learned

So far, we've done **manual, step-by-step EDA**:
- Loaded and inspected the breast cancer dataset (569 samples, 30 features)
- Checked for missing values, duplicates, and data types
- Analyzed feature correlations and identified top features by absolute correlation with target
- Ran t-tests to find statistically significant features (p < 0.05)
- Detected outliers using the IQR method and visualized them
- Confirmed class balance (relatively balanced, imbalance ratio ~1.05)
- Generated KDE plots, scatter plots, and a PCA projection
- Created ratio features (worst / mean) and dropped standard-error columns
- Applied percentile capping (1st–99th) and StandardScaler normalization
- Computed mutual information ranking of features

**Key outputs saved**:
- `data/engineered/breast_cancer_engineered.csv` — fully preprocessed dataset
- `artifacts/engineering/transformers.pkl` — scaler and metadata
- `artifacts/eda/mutual_info_ranking.csv` — feature importance

---


In [ ]:
# Define the EDA Agent — this mirrors the manual steps above but in an automated, repeatable way
class EDAAgent:
    """
    EDA Agent: Automates the entire EDA pipeline.
    
    This agent performs the same steps as the manual EDA above, but:
    - Logs each step with timestamps
    - Returns structured results (logs + metrics)
    - Can be reused and compared against the manual approach
    
    It's to compare:
    1. Manual approach (step-by-step human workflow)
    2. Agentic approach (automated pipeline with logs)
    
    Both should produce the same results, but the agent shows
    how automation + logging enables reproducibility and auditability.
    """
    
    def __init__(self, random_state=42, verbose=True):
        self.random_state = random_state
        self.verbose = verbose
        self.logs = []
        self.results = {}
    
    def log(self, message):
        """Record a timestamped log entry."""
        ts = time.strftime('%H:%M:%S')
        entry = f'[{ts}] {message}'
        self.logs.append(entry)
        if self.verbose:
            print(entry)
    
    def load_and_inspect(self, df, target_col):
        """Step 1: Load dataset and perform basic checks."""
        self.log(f'Loading dataset: shape {df.shape}')
        
        # Check for missing values
        missing_count = df.isna().sum().sum()
        if missing_count > 0:
            self.log(f'⚠ Found {missing_count} missing values')
        else:
            self.log('✓ No missing values detected')
        
        # Check for duplicates
        dup_count = df.duplicated().sum()
        self.log(f'Duplicates: {dup_count} rows')
        
        # Target distribution
        if target_col in df.columns:
            counts = df[target_col].value_counts()
            self.log(f'Target distribution: {dict(counts)}')
        
        self.results['dataset_shape'] = df.shape
        return df
    
    def analyze_correlations(self, X, y):
        """Step 2: Compute and rank feature correlations with target."""
        self.log('Computing feature-target correlations...')
        
        # Handle binary target
        if y.dtype == 'O' or y.dtype.name == 'category':
            uniq = list(y.unique())
            if len(uniq) == 2:
                y_bin = y.map({uniq[0]: 0, uniq[1]: 1})
            else:
                y_bin = y
        else:
            y_bin = y
        
        # Compute absolute correlations
        num_X = X.select_dtypes(include=[np.number])
        corr_with_target = num_X.corrwith(y_bin).abs().sort_values(ascending=False)
        
        top_features = corr_with_target.head(10)
        self.log(f'Top 10 correlated features: {list(top_features.index)}')
        
        self.results['top_correlated_features'] = top_features.to_dict()
        return y_bin
    
    def run_statistical_tests(self, X, y_bin):
        """Step 3: Perform t-tests to identify statistically significant features."""
        from scipy import stats
        
        self.log('Running t-tests for statistical significance...')
        
        class_0 = y_bin.unique()[0]
        class_1 = y_bin.unique()[1]
        
        X_class_0 = X[y_bin == class_0]
        X_class_1 = X[y_bin == class_1]
        
        sig_count = 0
        for col in X.select_dtypes(include=[np.number]).columns:
            t_stat, p_val = stats.ttest_ind(X_class_0[col].dropna(), X_class_1[col].dropna())
            if p_val < 0.05:
                sig_count += 1
        
        self.log(f'Found {sig_count} statistically significant features (p < 0.05)')
        self.results['significant_feature_count'] = sig_count
        return sig_count
    
    def detect_outliers(self, X):
        """Step 4: Identify outliers using IQR method."""
        self.log('Detecting outliers using IQR method...')
        
        outlier_count = 0
        for col in X.select_dtypes(include=[np.number]).columns:
            Q1, Q3 = X[col].quantile([0.25, 0.75])
            IQR = Q3 - Q1
            lower = Q1 - 1.5 * IQR
            upper = Q3 + 1.5 * IQR
            outliers = X[(X[col] < lower) | (X[col] > upper)]
            outlier_count += len(outliers)
        
        self.log(f'Total outlier instances detected: {outlier_count}')
        self.results['total_outliers'] = outlier_count
    
    def feature_engineering(self, X):
        """Step 5: Create ratio features and drop unnecessary columns."""
        self.log('Engineering features: creating ratios (worst / mean)...')
        
        X_fe = X.copy()
        
        # Create ratio features
        pairs = [
            ('radius_mean', 'radius_worst'),
            ('texture_mean', 'texture_worst'),
            ('perimeter_mean', 'perimeter_worst'),
            ('area_mean', 'area_worst'),
            ('smoothness_mean', 'smoothness_worst'),
            ('compactness_mean', 'compactness_worst'),
            ('concavity_mean', 'concavity_worst'),
            ('concave points_mean', 'concave points_worst'),
            ('symmetry_mean', 'symmetry_worst'),
            ('fractal_dimension_mean', 'fractal_dimension_worst'),
        ]
        
        ratio_count = 0
        for a, b in pairs:
            if a in X_fe.columns and b in X_fe.columns:
                with np.errstate(divide='ignore', invalid='ignore'):
                    X_fe[f'{b}_over_{a}'] = X_fe[b] / X_fe[a]
                ratio_count += 1
        
        self.log(f'Created {ratio_count} ratio features')
        
        # Drop standard-error columns
        se_cols = [c for c in X_fe.columns if str(c).endswith('_se')]
        X_fe = X_fe.drop(columns=se_cols, errors='ignore')
        self.log(f'Dropped {len(se_cols)} standard-error columns')
        
        self.results['engineered_shape'] = X_fe.shape
        return X_fe
    
    def scale_and_cap(self, X, lower_pct=0.01, upper_pct=0.99):
        """Step 6: Cap outliers at percentiles and apply StandardScaler."""
        self.log(f'Capping outliers at {lower_pct*100:.0f}th–{upper_pct*100:.0f}th percentiles...')
        
        X_cap = X.copy()
        for col in X.select_dtypes(include=[np.number]).columns:
            lo = X[col].quantile(lower_pct)
            hi = X[col].quantile(upper_pct)
            X_cap[col] = X_cap[col].clip(lo, hi)
        
        self.log('Applying StandardScaler to all numeric features...')
        scaler = StandardScaler()
        X_scaled = pd.DataFrame(scaler.fit_transform(X_cap), columns=X_cap.columns)
        
        self.log(f'Scaling complete. Final shape: {X_scaled.shape}')
        self.results['scaled_shape'] = X_scaled.shape
        
        return X_scaled, scaler
    
    def compute_feature_importance(self, X, y):
        """Step 7: Compute mutual information feature ranking."""
        self.log('Computing mutual information feature ranking...')
        
        # Impute missing values for MI computation
        imp = SimpleImputer(strategy='mean')
        X_imputed = pd.DataFrame(imp.fit_transform(X), columns=X.columns, index=X.index)
        
        # Handle binary target
        if y.dtype == 'O' or y.dtype.name == 'category':
            uniq = list(y.unique())
            if len(uniq) == 2:
                y_bin = y.map({uniq[0]: 0, uniq[1]: 1})
            else:
                y_bin = y
        else:
            y_bin = y
        
        # Compute MI
        mi = mutual_info_classif(X_imputed, y_bin, random_state=self.random_state)
        mi_series = pd.Series(mi, index=X.columns).sort_values(ascending=False)
        
        top_mi = mi_series.head(5).to_dict()
        self.log(f'Top 5 features by MI: {top_mi}')
        
        self.results['mi_ranking'] = mi_series.to_dict()
        return mi_series
    
    def run(self, df, target_col, X_cols=None):
        """
        Execute the full EDA pipeline.
        
        Parameters:
        - df: DataFrame with all features + target
        - target_col: name of the target column
        - X_cols: list of feature column names (if None, use all except target_col)
        
        Returns:
        - Dictionary with logs and results
        """
        self.log('='*60)
        self.log('EDA AGENT STARTED')
        self.log('='*60)
        
        # Step 1: Load and inspect
        df = self.load_and_inspect(df, target_col)
        
        # Split features and target
        y = df[target_col]
        X = df.drop(columns=[target_col])
        
        # Step 2: Analyze correlations
        y_bin = self.analyze_correlations(X, y)
        
        # Step 3: Statistical tests
        self.run_statistical_tests(X, y_bin)
        
        # Step 4: Outlier detection
        self.detect_outliers(X)
        
        # Step 5: Feature engineering
        X_fe = self.feature_engineering(X)
        
        # Step 6: Scale and cap
        X_scaled, scaler = self.scale_and_cap(X_fe)
        
        # Step 7: Feature importance
        self.compute_feature_importance(X_scaled, y)
        
        self.log('='*60)
        self.log('EDA AGENT FINISHED')
        self.log('='*60)
        
        return {
            'logs': self.logs,
            'results': self.results,
            'X_scaled': X_scaled,
            'y': y,
            'scaler': scaler
        }


---
## Running the EDA Agent — Automated Pipeline with Logs

Now let's instantiate the EDA Agent and run it on the same dataset.
The agent will perform **identical steps** as our manual EDA, but with:
- Timestamped logs for each operation
- Structured results that can be inspected and compared
- Reproducible, auditable output

Watch the agent's logs below to see how it progresses through the pipeline.


In [ ]:
# Instantiate the EDA Agent and run it
# Note: We're running it on the original df (same as our manual EDA above)

agent = EDAAgent(random_state=RANDOM_STATE, verbose=True)
agent_output = agent.run(df, target_col=TARGET_COL)

print('\n')
print('Agent execution complete.')


---
## Manual vs. Agentic EDA — Side-by-Side Comparison

Below we compare the manual approach (human-driven, step-by-step) with the agentic approach (automated pipeline).

**Key observations**:
- Both approaches should produce **identical results** (same dataset, same transformations)
- The agent provides **timestamped logs** showing exactly what happened and when
- The agent's **structured output** makes it easy to extract metrics and verify reproducibility
- For the professor: The logs show how automation eliminates manual errors and improves auditability


In [ ]:
# Display agent logs for full transparency
print('='*80)
print('AGENT EXECUTION LOG (Timestamped Events)')
print('='*80)
for i, log_entry in enumerate(agent_output['logs'], 1):
    print(f'{i:2d}. {log_entry}')

print('\n')
print('='*80)
print('AGENT RESULTS (Key Metrics Extracted)')
print('='*80)

results = agent_output['results']

print(f"\nDataset Shape: {results['dataset_shape']}")
print(f"Engineered Shape: {results['engineered_shape']}")
print(f"Scaled Shape: {results['scaled_shape']}")
print(f"Statistically Significant Features (p < 0.05): {results['significant_feature_count']}")
print(f"Total Outlier Instances Detected: {results['total_outliers']}")

print('\nTop Features by Correlation with Target:')
for feat, corr in list(results['top_correlated_features'].items())[:5]:
    print(f"  {feat}: {corr:.4f}")

print('\nTop Features by Mutual Information:')
mi_dict = results['mi_ranking']
for feat, mi_val in list(mi_dict.items())[:5]:
    print(f"  {feat}: {mi_val:.4f}")

print('\n' + '='*80)
print('COMPARISON: Manual vs. Agentic Approach')
print('='*80)
print(f"""
Manual EDA (Human-Driven):
  Flexible, exploratory approach
  Can inspect intermediate results
  Error-prone if repeated
  Hard to audit which steps were done
  Difficult to scale to new datasets

Agentic EDA (Automated Pipeline):
  Reproducible — same steps every time
  Auditable — full timestamped log of all operations
  Easy to reuse on new datasets
  Structured output for downstream analysis
  Can be integrated into larger workflows
  Less flexibility for ad-hoc exploration

CONCLUSION:
The agent produces the same EDA results as manual steps, but with reproducibility
and auditability. Perfect for consistent data processing in production pipelines.
""")


### Agentic-style Feature Engineering Demo

This short, non-redundant demo shows a one-line 'agentic' FE step that reuses the existing helper functions. It demonstrates how an agent would apply the baseline FE reproducibly (no new FEAgent class required).

In [ ]:
# Agentic-style FE demo (reuses existing helper functions)
X_fe_demo = add_ratio_features(X)
X_fe_demo = drop_se_columns(X_fe_demo)
X_fe_demo = cap_percentiles(X_fe_demo, lower=0.01, upper=0.99)
print('Agentic-style FE demo shape:', X_fe_demo.shape)
print('Sample engineered columns:', [c for c in X_fe_demo.columns if '_over_' in str(c)][:6])

## Baseline Feature Engineering
This cell applies our baseline feature engineering: it creates ratio features (worst / mean) for key measurements and removes all _se standard-error columns to reduce noise and dimensionality.

In [ ]:
# Create ratio features and drop *_se columns
def add_ratio_features(df):
    df = df.copy()
    pairs = [
        ('radius_mean','radius_worst'),
        ('texture_mean','texture_worst'),
        ('perimeter_mean','perimeter_worst'),
        ('area_mean','area_worst'),
        ('smoothness_mean','smoothness_worst'),
        ('compactness_mean','compactness_worst'),
        ('concavity_mean','concavity_worst'),
        ('concave points_mean','concave points_worst'),
        ('symmetry_mean','symmetry_worst'),
        ('fractal_dimension_mean','fractal_dimension_worst'),
    ]
    for a, b in pairs:
        if a in df.columns and b in df.columns:
            with np.errstate(divide='ignore', invalid='ignore'):
                df[f'{b}_over_{a}'] = df[b] / df[a]
    return df

def drop_se_columns(df):
    return df.drop(columns=[c for c in df.columns if str(c).endswith('_se')], errors='ignore')

X_fe = add_ratio_features(X)
X_fe = drop_se_columns(X_fe)
print("Feature-engineered shape:", X_fe.shape)


## Cap Percentiles, Scale, Save Engineered Data & Transformers
This cell caps outliers using percentile clipping, scales all features with StandardScaler, and then saves the engineered dataset and transformer metadata for consistent reuse across the project.

In [ ]:
def cap_percentiles(df, lower=0.01, upper=0.99):
    df = df.copy()
    for col in df.select_dtypes(include=[np.number]).columns:
        lo = df[col].quantile(lower)
        hi = df[col].quantile(upper)
        df[col] = df[col].clip(lo, hi)
    return df

lower_pct, upper_pct = 0.01, 0.99
X_cap = cap_percentiles(X_fe, lower=lower_pct, upper=upper_pct)

# Record clip bounds for metadata
clip_lower = {}
clip_upper = {}
for col in X_fe.select_dtypes(include=[np.number]).columns:
    clip_lower[col] = float(X_fe[col].quantile(lower_pct))
    clip_upper[col] = float(X_fe[col].quantile(upper_pct))

# Scale features
scaler = StandardScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(X_cap), columns=X_cap.columns)

fe_metadata = {
    "created_at": datetime.utcnow().isoformat() + "Z",
    "clip_percentiles": {"lower": lower_pct, "upper": upper_pct},
    "clip_bounds": {"lower": clip_lower, "upper": clip_upper},
    "feature_columns": list(X_scaled.columns),
    "transform": "cap_percentiles + StandardScaler",
}

out_df = X_scaled.copy()
out_df[TARGET_COL] = y.values

ENGINEERED_PATH = DATA_ENGINEERED / "breast_cancer_engineered.csv"
out_df.to_csv(ENGINEERED_PATH, index=False)

joblib.dump({'scaler': scaler, 'columns': list(X_scaled.columns), 'fe_metadata': fe_metadata},
            ARTIFACTS_ENG / 'transformers.pkl')

print("Saved engineered data to:", ENGINEERED_PATH)
print("Saved transformers to:", ARTIFACTS_ENG / "transformers.pkl")


## Mutual Information Ranking
This cell computes Mutual Information scores to measure how strongly each feature relates to the target, and saves a ranked list of the most informative features for later modeling.

In [ ]:
# Impute missing values with column mean
imp = SimpleImputer(strategy="mean")
X_imputed = pd.DataFrame(imp.fit_transform(X_scaled),
                         columns=X_scaled.columns, index=X_scaled.index)

# Select correct target (y_bin if present else y)
target = y_bin if 'y_bin' in globals() else y

# Compute mutual information
mi = mutual_info_classif(X_imputed, target, random_state=RANDOM_STATE)
mi_series = pd.Series(mi, index=X_imputed.columns).sort_values(ascending=False)

# Save ranking
mi_csv_path = ARTIFACTS_EDA / "mutual_info_ranking.csv"
mi_series.to_csv(mi_csv_path)
print("Saved mutual info ranking to:", mi_csv_path)

mi_series.head(10)


## Train/Test Split
This cell splits the processed data into training and testing sets (using stratification) so we can train models fairly and evaluate them on unseen data.

In [ ]:
# Use the same target as MI (binary if available)
target = y_bin if 'y_bin' in globals() else y

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, target,
    test_size=0.2,
    random_state=RANDOM_STATE,
    stratify=target,
)

print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)


## XGBoost + GridSearchCV Modeling
This cell trains an XGBoost model using GridSearchCV to find the best hyperparameters, then evaluates the final model on the test set using ROC-AUC, classification metrics, and a confusion matrix.

In [ ]:
# Define base XGBoost model
xgb = XGBClassifier(
    random_state=RANDOM_STATE,
    tree_method="hist",
    eval_metric="logloss",
)

# Hyperparameter grid (your earlier setup)
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'n_estimators': [400],
    'subsample': [0.8, 0.9, 1.0],
    'colsample_bytree': [0.8, 0.9, 1.0],
    'reg_lambda': [0.1, 1, 10],
    'min_child_weight': [1, 3, 5],
}

# Stratified K-fold CV
cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=RANDOM_STATE)

grid = GridSearchCV(
    estimator=xgb,
    param_grid=param_grid,
    scoring='roc_auc',
    n_jobs=-1,
    cv=cv,
    verbose=1,
)

grid.fit(X_train, y_train)

print("Best Parameters Found:", grid.best_params_)
print("Best ROC-AUC Score from CV:", grid.best_score_)

# Evaluate on test set
y_proba = grid.predict_proba(X_test)[:, 1]
test_auc = roc_auc_score(y_test, y_proba)
print("Test ROC-AUC:", test_auc)

y_pred = grid.predict(X_test)
print("\nClassification Report (Test Set):")
print(classification_report(y_test, y_pred))

# Confusion matrix
ConfusionMatrixDisplay.from_estimator(grid, X_test, y_test)
plt.title("Confusion Matrix — Best XGBoost Model")
plt.show()
